File: electron/jwpub-crypto.ts
import { createDecipheriv, createHash } from 'node:crypto';
import { inflate } from 'pako';
import type { Database } from 'sql.js';

const XOR_KEY = Buffer.from('11cbb5587e32846d4c26790c633da289f66fe5842a3a585ce1bc3a294af5ada7', 'hex');

export function deriveKeyIv(db: Database): Buffer {
  const row = db.exec(
    'SELECT MepsLanguageIndex, Symbol, Year, IssueTagNumber FROM Publication LIMIT 1',
  )[0]?.values?.[0];
  if (!row) throw new Error('Tabela Publication vazia');

  const [lang, symbol, year, issue] = row.map(String);
  const parts = [lang, symbol, year];
  if (issue && issue !== '0' && Number(issue) !== 0) {
    parts.push(issue);
  }

  const hash = createHash('sha256').update(parts.join('_')).digest();
  const keyIv = Buffer.alloc(32);
  for (let i = 0; i < 32; i++) keyIv[i] = hash[i] ^ XOR_KEY[i];
  return keyIv;
}

export function decryptContent(keyIv: Buffer, encrypted: Uint8Array | Buffer): string {
  const buf = Buffer.from(encrypted);
  const decipher = createDecipheriv('aes-128-cbc', keyIv.subarray(0, 16), keyIv.subarray(16, 32));
  const decrypted = Buffer.concat([decipher.update(buf), decipher.final()]);
  return inflate(decrypted, { to: 'string' });
}

File: tnme-biblia/android/app/src/main/java/com/jcs/tnme/JwpubCrypto.java
package com.jcs.tnme;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.zip.Inflater;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

final class JwpubCrypto {
    private static final byte[] XOR_KEY = hexToBytes(
        "11cbb5587e32846d4c26790c633da289f66fe5842a3a585ce1bc3a294af5ada7");

    private JwpubCrypto() {}

    static byte[] deriveKeyIv(String lang, String symbol, String year, String issue) {
        StringBuilder joined = new StringBuilder();
        joined.append(lang).append('_').append(symbol).append('_').append(year);
        if (issue != null && issue.length() > 0 && !"0".equals(issue)) {
            try {
                if (Integer.parseInt(issue) != 0) {
                    joined.append('_').append(issue);
                }
            } catch (NumberFormatException ignored) {
                joined.append('_').append(issue);
            }
        }

        byte[] hash = sha256(joined.toString().getBytes(StandardCharsets.UTF_8));
        byte[] keyIv = new byte[32];
        for (int i = 0; i < 32; i++) {
            keyIv[i] = (byte) (hash[i] ^ XOR_KEY[i]);
        }
        return keyIv;
    }

    static String decryptContent(byte[] keyIv, byte[] encrypted) throws Exception {
        SecretKeySpec keySpec = new SecretKeySpec(keyIv, 0, 16, "AES");
        IvParameterSpec ivSpec = new IvParameterSpec(keyIv, 16, 16);
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);
        byte[] decrypted = cipher.doFinal(encrypted);

        Inflater inflater = new Inflater();
        inflater.setInput(decrypted);
        byte[] buffer = new byte[Math.max(decrypted.length * 4, 8192)];
        int length = inflater.inflate(buffer);
        inflater.end();
        return new String(buffer, 0, length, StandardCharsets.UTF_8);
    }

    private static byte[] sha256(byte[] input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return digest.digest(input);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] hexToBytes(String hex) {
        int len = hex.length();
        byte[] out = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            out[i / 2] = (byte) Integer.parseInt(hex.substring(i, i + 2), 16);
        }
        return out;
    }
}
